jscomp/syntax/reactjs_jsx_ppx_v3.ml

```diff
@@ -0,0 +1,1092 @@
+
+  let transformComponentSignature _mapper signature returnSignatures = match signature with
+  | ({
+      psig_loc;
+      psig_desc = Psig_value ({
+        pval_name = { txt = fnName };
+        pval_attributes;
+        pval_type;
+      } as psig_desc)
+    } as psig) ->
+    (match List.filter hasAttr pval_attributes with
+    | [] -> signature :: returnSignatures
+    | [_] ->
+    let rec getPropTypes types ({ptyp_loc; ptyp_desc} as fullType) =
+      (match ptyp_desc with
+      | Ptyp_arrow (name, type_, ({ptyp_desc = Ptyp_arrow _} as rest)) when isOptional name || isLabelled name ->
+        getPropTypes ((name, ptyp_loc, type_)::types) rest
+#if OCAML_VERSION =~ ">4.03.0" then
+      | Ptyp_arrow (Nolabel, _type, rest) ->
+#else
+      | Ptyp_arrow ("", _type, rest) ->
+#end
+        getPropTypes types rest
+      | Ptyp_arrow (name, type_, returnValue) when isOptional name || isLabelled name ->
+        (returnValue, (name, returnValue.ptyp_loc, type_)::types)
+      | _ -> (fullType, types))
+    in
+    let (innerType, propTypes) = getPropTypes [] pval_type in
+    let namedTypeList = List.fold_left argToConcreteType [] propTypes in
+    let pluckLabelAndLoc (label, loc, type_) = (label, None, loc, Some type_) in
+    let retPropsType = makePropsType ~loc:psig_loc namedTypeList in
+    let externalPropsDecl = makePropsExternalSig fnName psig_loc ((
+      optional "key",
+      None,
+      psig_loc,
+      Some(keyType psig_loc)
+    ) :: List.map pluckLabelAndLoc propTypes) retPropsType in
+        (* can't be an arrow because it will defensively uncurry *)
+    let newExternalType = Ptyp_constr (
+      {loc = psig_loc; txt = Ldot ((Lident "React"), "componentLike")},
+      [retPropsType; innerType]
+    ) in
+    let newStructure = {
+      psig with psig_desc = Psig_value {
+        psig_desc with pval_type = {
+          pval_type with ptyp_desc = newExternalType;
+        };
+        pval_attributes = List.filter otherAttrsPure pval_attributes;
+      }
+    } in
+    externalPropsDecl :: newStructure :: returnSignatures
+    | _ -> raise (Invalid_argument "Only one react.component call can exist on a component at one time"))
+  | signature -> signature :: returnSignatures in
+
+  let reactComponentSignatureTransform mapper signatures =
+  List.fold_right (transformComponentSignature mapper) signatures [] in
+
+
+  let transformJsxCall mapper callExpression callArguments attrs =
+    (match callExpression.pexp_desc with
+     | Pexp_ident caller ->
+       (match caller with
+        | {txt = Lident "createElement"} ->
+          raise (Invalid_argument "JSX: `createElement` should be preceeded by a module name.")
+
+        (* Foo.createElement(~prop1=foo, ~prop2=bar, ~children=[], ()) *)
+        | {loc; txt = Ldot (modulePath, ("createElement" | "make"))} ->
+          (match !jsxVersion with
+          | Some 2 -> transformUppercaseCall modulePath mapper loc attrs callExpression callArguments
+          | None
```

Comment:
v2 vs v3 should actually be another cppo statement here and a few lines below. Is there a good/easy way to add another conditional variable? That way we can fully replace the other copied files.